"Add Variation" Button

ThomazFB opened the pull request:
Summary
==========
Address issue #3205 with two changes:

1 - Adapting the Variations `ProductDetailCardBuilder` method to now present the `Add Variation` text instead of the previous `This product has no variation yet` description.

2 - Adding to `VariationListFragment`a new kind of `ActionableEmptyView` that presents the `Add Variation` view and action instead of suggesting the user go to the web.

⚠️ The `Add Variation` button behavior will be implemented on upcoming PRs

How to Test
==========

New behavior (only available when on debug mode)
----------
| Click on Products section and select a Variable Product | Click on `Add Variations` and make sure no view is called instead of the old web redirection suggestion |
| ------------- | ------------- |
| ![Screenshot_1614251390](https://user-images.githubusercontent.com/5920403/109145190-ecd61f00-7740-11eb-8eb5-0a7a0158924a.png) | ![Screenshot_1614247056](https://user-images.githubusercontent.com/5920403/109145672-8e5d7080-7741-11eb-9b1a-f987efb38b82.png) |
Old behavior
----------
| Click on Products section and select a Variable Product | Click on `Variations` | Make sure presented View informs the user about adding variations only available on the web |
| ------------- | ------------- | ------------- |
| ![Screenshot_1614251390](https://user-images.githubusercontent.com/5920403/109145190-ecd61f00-7740-11eb-8eb5-0a7a0158924a.png) | ![Screenshot_1614251359](https://user-images.githubusercontent.com/5920403/109145594-7128a200-7741-11eb-925c-3d20a05a3cdf.png) | ![Screenshot_1614251362](https://user-images.githubusercontent.com/5920403/109145623-7a197380-7741-11eb-8081-fb7562e1837d.png) |

Update release notes:

- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ThomazFB committed:
Change string from no variation to add variation action

ThomazFB committed:
Update empty variation list string to add variation flow info

ThomazFB committed:
Update empty variation list button to add variation flow info

ThomazFB committed:
Empty variation list UI adjustments

ThomazFB committed:
Adjust ActionableEmptyView to support button behavior setup

ThomazFB committed:
Make Add Variation button visible and available

ThomazFB committed:
Merge branch 'develop' into issue/3205-add-attribute-button

ThomazFB committed:
Reorganize string to avoid side effects on UI

ThomazFB committed:
Revert the empty view in order to support feature flag switching

ThomazFB committed:
Control variations empty view scenario through Feature Flag

ThomazFB committed:
Fix Lint issues

ThomazFB committed:
Change onClickListener target for VariationListFragment

ThomazFB committed:
Improve variations ComplexProperty to not show the title

ThomazFB committed:
Add feature flag switching to Variations Complex Property creation

nbradbury commented:
@ThomazFB This is really a design issue, but it seems awkward to tap "Add variations" in the product detail screen and be taken to another screen where you have to tap again to add a variation. Also, the button on the second screen says "Add attribute" - shouldn't that be "Add variation?"

![1](https://user-images.githubusercontent.com/3903757/109171865-2c583780-7750-11eb-83af-b8f0488ee4fc.png)

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/VariationListFragment.kt:
```diff
@@ -232,4 +241,13 @@ class VariationListFragment : BaseFragment(R.layout.fragment_variation_list),
         adapter.setVariationList(variations)
     }
+
+    private fun ActionableEmptyView.handleEmptyViewVisibility(shouldBeVisible: Boolean, showButton: Boolean) {
```
Would it make more sense to move this into `ActionableEmptyView`? Also, I think we'd want to check the visibility before running the animation, eg:

```
if (shouldBeVisible && !isVisible) {
     WooAnimUtils.fadeIn(this)
     showButton(showButton)
} else if (!shouldBeVisible && isVisible) {
     WooAnimUtils.fadeOut(this)
 }
```

ThomazFB replied:
Agreed! Much better this way, I've submitted the changes right here: https://github.com/woocommerce/woocommerce-android/pull/3620/commits/9bbdc18da6eed3bfc9bf11c363aa247397be964c

ThomazFB committed:
Fix incorrect text on Empty VariationListFragment

ThomazFB committed:
Refactor visibility changing methods for ActionableEmptyView

ThomazFB committed:
Fix Lint errors

ThomazFB commented:
> This is really a design issue, but it seems awkward to tap "Add variations" in the product detail screen and be taken to another screen where you have to tap again to add a variation. 

Yeah, I found that to be a little bit odd too, it does make sense to prompt the `Add your first variation` view when we delete the last Variation available to a given product on the list, but selecting the `Add Attribute` from the Product Details view should direct the user immediately to the Variation creation flow, that was a nice catch @nbradbury! 

@Garance91540 @adamzelinski what are your thoughts about this point Nick suggested?

CC: @Ecarrion 

> Also, the button on the second screen says "Add attribute" - shouldn't that be "Add variation?"

Oh, sorry about that 🤦‍♂️, it should be "Add variation", I've fixed that here: https://github.com/woocommerce/woocommerce-android/pull/3620/commits/ea58518d3e0ab2961661c6cde37945fe7a2a72c6

Garance91540 commented:
Hi there, in my initial designs, when selecting "Add variations" on the product page, users where directly taken to the ["add attribute" screen](https://d.pr/i/k1nHQY) as you suggested. But because of the way the navigation is set up on iOS, we had to add back that empty state "[add variation](https://d.pr/i/CC72Tf)" screen. So if it's possible to skip it in Android, that's awesome! 

So we should only display this screen if a user has deleted all variations, and you are right @nbradbury the CTA should be "Add Variation" and not "Add Attribute"

Ecarrion commented:
> So we should only display this screen if a user has deleted all variations, and you are right @nbradbury the CTA should be "Add Variation" and not "Add Attribute"

FYI: It should also be shown if the user has deleted all attributes, as core won't delete the variations for you!

ThomazFB committed:
Merge branch 'develop' into issue/3205-add-attribute-button

ThomazFB committed:
Change ProductDetailCardBuilder "Add Variation" button behavior

ThomazFB committed:
Update TODO text for VariationListFragment

ThomazFB commented:
Hey @nbradbury, the PR is ready for another round, but since now we changed the behavior of the `Add Variation` button at Product Details View as we discussed here, and currently we don't support deleting variations from the app, it's worth noting that the new `Add your first variation` view prompt at the Variation List View is no longer reachable and it will be back only after we fully support managing the Product Variation list, nevertheless I'm leaving here the code for that view since it's behind a feature flag and it's correctly functional for upcoming PRs.

I've changed the PR description to correctly represent the currently available flow for testing, which is much smaller now.

Also, thanks @Garance91540 and @Ecarrion for the awesome feedbacks and tips!

nbradbury reviewed (approved):
:shipit:
